pkg/controller/modeler/modeler.go

```diff
@@ -0,0 +1,227 @@
+
+	"github.com/sethvargo/go-limiter"
+	"go.uber.org/zap"
+)
+
+// Controller is a controller for the modeler service.
+type Controller struct {
+	config  *config.Modeler
+	db      *database.Database
+	h       *render.Renderer
+	limiter limiter.Store
+	logger  *zap.SugaredLogger
+}
+
+// New creates a new modeler controller.
+func New(ctx context.Context, config *config.Modeler, db *database.Database, limiter limiter.Store, h *render.Renderer) *Controller {
+	logger := logging.FromContext(ctx).Named("modeler")
+
+	return &Controller{
+		config:  config,
+		db:      db,
+		h:       h,
+		limiter: limiter,
+		logger:  logger,
+	}
+}
+
+// HandleModel accepts an HTTP trigger and re-generates the models.
+func (c *Controller) HandleModel() http.Handler {
+	logger := c.logger.Named("HandleModel")
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		if err := c.rebuildModels(ctx); err != nil {
+			logger.Errorw("failed to build models", "error", err)
+			c.h.JSON500(w, err)
+			return
+		}
+
+		c.h.RenderJSON(w, http.StatusOK, nil)
+	})
+}
+
+// rebuildModels iterates over all models with abuse prevention enabled and
+// calculates the new limits.
+func (c *Controller) rebuildModels(ctx context.Context) error {
+	logger := c.logger.Named("rebuildModels")
+	db := c.db.RawDB()
+
+	// Get all realm IDs in a single operation so we can iterate realm-by-realm to
+	// avoid a full table lock during stats calculation.
+	var ids []uint64
+	if err := db.
+		Model(&database.Realm{}).
+		Where("abuse_prevention_enabled IS true").
+		Pluck("id", &ids).
+		Error; err != nil {
+		return fmt.Errorf("failed to fetch ids: %w", err)
+	}
+	logger.Debugw("building models", "count", len(ids))
+
+	// Process all models.
+	var merr *multierror.Error
+	for _, id := range ids {
+		if err := c.rebuildModel(ctx, id); err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("failed to update realm %d: %w", id, err))
+		}
+	}
+
+	return merr.ErrorOrNil()
+}
+
+// rebuildModel rebuilds the model for a single model.
+func (c *Controller) rebuildModel(ctx context.Context, id uint64) error {
+	logger := c.logger.Named("rebuildModel").With("id", id)
+	db := c.db.RawDB()
+
+	// Get 21 days of historical data for the realm.
+	var ys []float64
+	if err := db.
+		Model(&database.RealmStats{}).
+		Where("realm_id = ?", id).
+		Order("date DESC").
+		Limit(21).
+		Pluck("codes_issued", &ys).
+		Error; err != nil {
+		return fmt.Errorf("failed to get stats: %w", err)
+	}
+
+	// Require some reasonable number of days of history before attempting to
+	// build a model.
+	if l := len(ys); l < 14 {
+		logger.Warnw("skipping, not enough data", "points", l)
+		return nil
+	}
+
+	// Exclude the most recent record. Depending on timezones, the "day" might not
+	// be over at 00:00 UTC, and we don't want to generate a partial model.
+	ys = ys[:len(ys)-1]
+
+	// Reverse the list - it came in reversed because we sorted by date DESC, but
```

Comment:
Wanna just ask for the results ascending above? I am not enough of an SQL hacker to know if there's a query that does what you want there.

Reply:
We're building the regression model based on the past 21 days of data, so we need the SQL query to be "the last 21 records" (order by date DESC). I don't think there's a SQL way to ORDER BY and take the _last_ elements.